src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -386,34 +410,111 @@ public override int GetHashCode()
 
         internal override void AddNullableTransforms(ArrayBuilder<byte> transforms)
         {
-            ElementTypeWithAnnotations.AddNullableTransforms(transforms);
+            var elementType = ElementTypeWithAnnotations;
+            while (elementType.Type is ArrayTypeSymbol arrayType)
+            {
+                elementType.AddNullableTransformShallow(transforms);
+                elementType = arrayType.ElementTypeWithAnnotations;
+            }
+
+            elementType.AddNullableTransforms(transforms);
         }
 
-        internal override bool ApplyNullableTransforms(byte defaultTransformFlag, ImmutableArray<byte> transforms, ref int position, out TypeSymbol result)
+        internal override TypeSymbol ApplyNullableTransforms(NullableTransformStream stream)
         {
-            TypeWithAnnotations oldElementType = ElementTypeWithAnnotations;
-            TypeWithAnnotations newElementType;
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach.
+            var builder = ArrayBuilder<(ArrayTypeSymbol array, byte? transform)>.GetInstance();
 
-            if (!oldElementType.ApplyNullableTransforms(defaultTransformFlag, transforms, ref position, out newElementType))
+            // First build up the state for each of the ArrayTypeSymbol in the chain of arrays.
+            for (var array = this; array is object; array = array.ElementType as ArrayTypeSymbol)
             {
-                result = this;
-                return false;
+                builder.Push((array, stream.GetNextTransform()));
             }
 
-            result = WithElementType(newElementType);
-            return true;
+            // Unwind and rebuild the array
+            var finalArray = builder.Peek().array.ElementType.ApplyNullableTransforms(stream);
+            do
+            {
+                var state = builder.Pop();
```

Comment:
>var state [](start = 16, length = 9)

Please consider deconstructing:
```
var (array, transform) = builder.Pop();
```

Reply:
Hadn' tthought o f that.

---
In reply to: [275071261](https://github.com/dotnet/roslyn/pull/34891#discussion_r275071261) [](ancestors = 275071261)